crossplane/crossplane
# stacks: sync image pull secrets to the host

### Description of your changes
 When a StackInstall or Stack (controller template) include imagePullSecrets, these secrets must be copied to the host where the install job (for StackInstalls) and the deployment (for Stacks) are run.
    
This PR adds to the preparation of jobs and deployments a process that:
    
* Creates secure, length restricted, alternate names for the tenant secret to use on the host. 
 Predictable secret names could be abused by stacks in other namespaces.
* References those names in the host job or deployment
* Creates the job or deployment (per usual)
* Copies the secrets from the tenant into their host mapped names, using the job or deployment as an owner reference for garbage collection
 

Fixes #1317 

#### Learnings

* `client.Create` sets the `UID` on the object immediately on a successful REST response
* Job and Deployment `spec.template.spec.imagePullSecrets` fields are immutable

### How has this code been tested?

Unit tests walk the `hosted.SyncImagePullSecrets` flow, with the names of local and host secrets being plugged in.

In local testing, using `cluster/local/hosted_test.sh`:
* push the `app-wordpress` image to a private repo (on Docker hub, for example)
* create a `regcred` secret in the tenant: 
  `kubectl create secret docker-registry regcred --docker-server=docker.io --docker-username=displague --docker-password=... --docker-email=marques@example.com`
* create a StackInstall in the tenant that requires this credential as an image pull secret
* watch the Job and Deployment succeed in the `crossplane-host-controller` namespace
* notice that the tenant secret has been copied into two new secrets in the `crossplane-host-controller` namespace
* notice that the `Stack` and `StackInstall` have both successfully reconciled

```
apiVersion: stacks.crossplane.io/v1alpha1
kind: StackInstall
metadata:
  name: "app-wordpress"
  namespace: "default"
spec:
  package: displague/app-wordpress:master
  source: index.docker.io
  imagePullSecrets:
  - name: regcred
  imagePullPolicy: Always
```

### Checklist

I have:
- [ ] Updated any relevant [documentation] and [examples].
- [ ] Reported all new error conditions into the log or as an event, as
  appropriate.

For more about what we believe makes a pull request complete, see our
[definition of done].

[documentation]: https://github.com/crossplane/crossplane/tree/master/docs
[examples]: https://github.com/crossplane/crossplane/tree/master/cluster/examples
[definition of done]: https://github.com/crossplane/crossplane/tree/master/design/one-pager-definition-of-done.md
